File: code/ch12/12.4.1.GetMoneyAmount.py
import sys


class Solution:
    def getMoneyAmount(self, n: int) -> int:
        def cost(low: int, high: int) -> int:
            if low >= high:
                return 0
            res = sys.maxsize
            for i in range(low, high + 1):
                tmp = i + max(
                    cost(low, i - 1), cost(i + 1, high)
                )
                res = min(res, tmp)
            return res

        return cost(1, n)

File: code/ch12/12.4.1.GetMoneyAmount.cpp
class Solution {
private:
    int cost(int low, int high) {
        if (low >= high) return 0;
        
        int res = INT_MAX;
        for (int i = low; i < high + 1; i++) {
            int tmp = i + max(cost(low, i - 1), cost(i + 1, high));
            res = min(res, tmp);
        }
        
        return res;
    }
public:
    int getMoneyAmount(int n) {
        return cost(1, n);
    }
};
